File: tools/clod-poc/src/world_source/biome_region_field.ts
import { resolveIslandShapeConfig, sampleIslandMask, type IslandMaskSample, type IslandShapeConfig } from "./island_shape.js";

export const BIOME_IDS = {
  meadows: 0,
  forest: 1,
  swamp: 2,
  mountain: 3,
  plains: 4,
  coast: 5,
  ocean: 6,
} as const;

export const BIOME_REGION_CELL_M = 420;
export const BIOME_OCEAN_HEIGHT_MARGIN_M = 1.5;
export const BIOME_OCEAN_ISLAND_MASK_MAX = 0.08;
export const BIOME_COAST_HEIGHT_BAND_M = 4;
export const BIOME_COAST_SHORE_DISTANCE_M = 42;
export const BIOME_MOUNTAIN_HEIGHT_ABOVE_SEA_M = 68;
export const BIOME_SWAMP_HEIGHT_ABOVE_SEA_M = 8;
export const BIOME_SWAMP_NOISE_MAX = 0.42;
export const BIOME_PLAINS_DISTANCE_MIN = 0.72;
export const BIOME_PLAINS_NOISE_MIN = 0.58;
export const BIOME_FOREST_NOISE_MIN = 0.46;

export type BiomeId = typeof BIOME_IDS[keyof typeof BIOME_IDS];

export interface BiomeRegionSample {
  biome: BiomeId;
  regionNoise: number;
  islandDistanceT: number;
}

export interface BiomeRegionFieldOptions {
  seed: number;
  seaLevel: number;
  /** Must stay equal to BIOME_REGION_CELL_M until WGSL accepts this as a uniform. */
  regionCellM?: number;
  islandShape?: Partial<IslandShapeConfig>;
}

export interface BiomeRegionClassifyInput {
  x: number;
  z: number;
  height: number;
  seed: number;
  seaLevel: number;
  regionCellM: number;
  islandRadiusM: number;
  island: IslandMaskSample;
}

function mix32(value: number): number {
  let mixed = value >>> 0;
  mixed ^= mixed >>> 16;
  mixed = Math.imul(mixed, 0x7feb352d);
  mixed ^= mixed >>> 15;
  mixed = Math.imul(mixed, 0x846ca68b);
  mixed ^= mixed >>> 16;
  return mixed >>> 0;
}

function resolveRegionCellM(value: unknown): number {
  if (value === undefined) return BIOME_REGION_CELL_M;
  if (typeof value !== "number" || !Number.isFinite(value)) return BIOME_REGION_CELL_M;
  if (value !== BIOME_REGION_CELL_M) {
    throw new Error(`BiomeRegionField regionCellM must be ${BIOME_REGION_CELL_M} to match the GPU shader; got ${value}`);
  }
  return BIOME_REGION_CELL_M;
}

export function pcg2d(x: number, z: number, seed: number): number {
  const value = (seed >>> 0)
    ^ Math.imul(x | 0, 0x1f123bb5)
    ^ Math.imul(z | 0, 0x5f356495);
  return mix32(value) / 0x100000000;
}

function smooth01(value: number): number {
  const t = Math.min(1, Math.max(0, value));
  return t * t * (3 - 2 * t);
}

export function biomeRegionNoise(x: number, z: number, cellM: number, seed: number): number {
  const safeCellM = cellM === BIOME_REGION_CELL_M ? BIOME_REGION_CELL_M : resolveRegionCellM(cellM);
  const gx = x / safeCellM;
  const gz = z / safeCellM;
  const x0 = Math.floor(gx);
  const z0 = Math.floor(gz);
  const tx = smooth01(gx - x0);
  const tz = smooth01(gz - z0);
  const a = pcg2d(x0, z0, seed);
  const b = pcg2d(x0 + 1, z0, seed);
  const c = pcg2d(x0, z0 + 1, seed);
  const d = pcg2d(x0 + 1, z0 + 1, seed);
  return a + (b - a) * tx + (c - a) * tz + (a - b - c + d) * tx * tz;
}

export function classifyBiomeRegion(input: BiomeRegionClassifyInput): BiomeRegionSample {
  const { x, z, height, seed, seaLevel, regionCellM, islandRadiusM, island } = input;
  if (height < seaLevel - BIOME_OCEAN_HEIGHT_MARGIN_M || island.mask < BIOME_OCEAN_ISLAND_MASK_MAX) {
    return { biome: BIOME_IDS.ocean, regionNoise: 0, islandDistanceT: 0 };
  }
  if (Math.abs(height - seaLevel) < BIOME_COAST_HEIGHT_BAND_M || island.shoreDistanceM < BIOME_COAST_SHORE_DISTANCE_M) {
    return { biome: BIOME_IDS.coast, regionNoise: 0, islandDistanceT: 0 };
  }

  const n = biomeRegionNoise(x, z, regionCellM, seed + 711);
  const centerDistance = Math.hypot(x - island.nearestCenterX, z - island.nearestCenterZ);
  const distanceT = Math.min(1, Math.max(0, centerDistance / Math.max(1, islandRadiusM)));

  if (height >= seaLevel + BIOME_MOUNTAIN_HEIGHT_ABOVE_SEA_M) return { biome: BIOME_IDS.mountain, regionNoise: n, islandDistanceT: distanceT };
  if (height <= seaLevel + BIOME_SWAMP_HEIGHT_ABOVE_SEA_M && n < BIOME_SWAMP_NOISE_MAX) return { biome: BIOME_IDS.swamp, regionNoise: n, islandDistanceT: distanceT };
  if (distanceT > BIOME_PLAINS_DISTANCE_MIN && n > BIOME_PLAINS_NOISE_MIN) return { biome: BIOME_IDS.plains, regionNoise: n, islandDistanceT: distanceT };
  if (n > BIOME_FOREST_NOISE_MIN) return { biome: BIOME_IDS.forest, regionNoise: n, islandDistanceT: distanceT };
  return { biome: BIOME_IDS.meadows, regionNoise: n, islandDistanceT: distanceT };
}

export class BiomeRegionField {
  readonly seed: number;
  readonly seaLevel: number;
  readonly regionCellM: number;
  readonly islandShape: IslandShapeConfig;

  constructor(options: BiomeRegionFieldOptions) {
    this.seed = Number.isFinite(options.seed) ? Math.floor(options.seed) : 0;
    this.seaLevel = Number.isFinite(options.seaLevel) ? options.seaLevel : 18;
    this.regionCellM = resolveRegionCellM(options.regionCellM);
    this.islandShape = resolveIslandShapeConfig({
      ...options.islandShape,
      seed: options.islandShape?.seed ?? this.seed,
      seaLevel: this.seaLevel,
    });
  }

  sample(x: number, z: number, height: number): BiomeRegionSample {
    return classifyBiomeRegion({
      x,
      z,
      height,
      seed: this.seed,
      seaLevel: this.seaLevel,
      regionCellM: this.regionCellM,
      islandRadiusM: this.islandShape.radiusM,
      island: sampleIslandMask(x, z, this.islandShape),
    });
  }
}

File: src/world/source/biome_region_field.rs
use bevy::prelude::Resource;
use serde::{Deserialize, Serialize};

use super::biome_region_contract::{
    BIOME_REGION_CELL_M, BIOME_REGION_CONTRACT, BiomeRegionContract,
};
use super::island_shape::{IslandMaskSample, IslandShapeConfig, sample_island_mask};
use super::noise::smooth01;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[repr(u8)]
pub enum BiomeId {
    Meadows = 0,
    Forest = 1,
    Swamp = 2,
    Mountain = 3,
    Plains = 4,
    Coast = 5,
    Ocean = 6,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct BiomeRegionSample {
    pub biome: BiomeId,
    pub region_noise: f32,
    pub island_distance_t: f32,
}

#[derive(Resource, Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct BiomeRegionField {
    pub seed: i32,
    pub sea_level: f32,
    pub region_cell_m: f32,
    pub island_shape: IslandShapeConfig,
    pub contract: BiomeRegionContract,
}

#[derive(Debug, Clone, Copy)]
pub struct BiomeRegionClassifyInput {
    pub x: f32,
    pub z: f32,
    pub height: f32,
    pub seed: i32,
    pub sea_level: f32,
    pub region_cell_m: f32,
    pub island_radius_m: f32,
    pub island: IslandMaskSample,
    pub contract: BiomeRegionContract,
}

impl BiomeId {
    pub fn layer_index(self) -> u32 {
        self as u32
    }
}

impl BiomeRegionField {
    pub fn new(seed: i32, sea_level: f32, island_shape: IslandShapeConfig) -> Self {
        Self {
            seed,
            sea_level,
            region_cell_m: BIOME_REGION_CELL_M,
            island_shape,
            contract: BIOME_REGION_CONTRACT,
        }
    }

    pub fn with_region_cell_m(mut self, region_cell_m: f32) -> Self {
        self.region_cell_m = resolve_region_cell_m(region_cell_m, self.contract);
        self
    }

    pub fn with_contract(mut self, contract: BiomeRegionContract) -> Self {
        self.contract = contract;
        self.region_cell_m = resolve_region_cell_m(contract.region_cell_m, contract);
        self
    }

    pub fn sample(&self, x: f32, z: f32, height: f32) -> BiomeRegionSample {
        let island = sample_island_mask(x, z, &self.island_shape);
        classify_biome_region(BiomeRegionClassifyInput {
            x,
            z,
            height,
            seed: self.seed,
            sea_level: self.sea_level,
            region_cell_m: self.region_cell_m,
            island_radius_m: self.island_shape.radius_m.max(1.0),
            island,
            contract: self.contract,
        })
    }
}

impl Default for BiomeRegionField {
    fn default() -> Self {
        Self::new(0, 18.0, IslandShapeConfig::default())
    }
}

fn resolve_region_cell_m(region_cell_m: f32, contract: BiomeRegionContract) -> f32 {
    if !region_cell_m.is_finite() {
        return contract.region_cell_m;
    }
    if (region_cell_m - contract.region_cell_m).abs() > f32::EPSILON {
        panic!(
            "BiomeRegionField region_cell_m must be {} to match the shared GPU contract; got {}",
            contract.region_cell_m, region_cell_m,
        );
    }
    contract.region_cell_m
}

fn mix32(value: u32) -> u32 {
    let mut mixed = value;
    mixed ^= mixed >> 16;
    mixed = mixed.wrapping_mul(0x7feb_352d);
    mixed ^= mixed >> 15;
    mixed = mixed.wrapping_mul(0x846c_a68b);
    mixed ^= mixed >> 16;
    mixed
}

pub fn pcg2d(x: i32, z: i32, seed: i32) -> f32 {
    let value =
        (seed as u32) ^ (x as u32).wrapping_mul(0x1f12_3bb5) ^ (z as u32).wrapping_mul(0x5f35_6495);
    mix32(value) as f32 / 4_294_967_296.0
}

pub fn biome_region_noise(
    x: f32,
    z: f32,
    cell_m: f32,
    seed: i32,
    contract: BiomeRegionContract,
) -> f32 {
    let cell_m = resolve_region_cell_m(cell_m, contract);
    let gx = x / cell_m;
    let gz = z / cell_m;
    let x0 = gx.floor() as i32;
    let z0 = gz.floor() as i32;
    let tx = smooth01(gx - x0 as f32);
    let tz = smooth01(gz - z0 as f32);
    let a = pcg2d(x0, z0, seed);
    let b = pcg2d(x0.wrapping_add(1), z0, seed);
    let c = pcg2d(x0, z0.wrapping_add(1), seed);
    let d = pcg2d(x0.wrapping_add(1), z0.wrapping_add(1), seed);
    a + (b - a) * tx + (c - a) * tz + (a - b - c + d) * tx * tz
}

pub fn classify_biome_region(input: BiomeRegionClassifyInput) -> BiomeRegionSample {
    if input.height < input.sea_level - input.contract.ocean_height_margin_m
        || input.island.mask < input.contract.ocean_island_mask_max
    {
        return BiomeRegionSample {
            biome: BiomeId::Ocean,
            region_noise: 0.0,
            island_distance_t: 0.0,
        };
    }
    if (input.height - input.sea_level).abs() < input.contract.coast_height_band_m
        || input.island.shore_distance_m < input.contract.coast_shore_distance_m
    {
        return BiomeRegionSample {
            biome: BiomeId::Coast,
            region_noise: 0.0,
            island_distance_t: 0.0,
        };
    }

    let n = biome_region_noise(
        input.x,
        input.z,
        input.region_cell_m,
        input.seed.wrapping_add(711),
        input.contract,
    );
    let center_distance =
        (input.x - input.island.nearest_center_x).hypot(input.z - input.island.nearest_center_z);
    let distance_t = (center_distance / input.island_radius_m.max(1.0)).clamp(0.0, 1.0);

    let biome = if input.height >= input.sea_level + input.contract.mountain_height_above_sea_m {
        BiomeId::Mountain
    } else if input.height <= input.sea_level + input.contract.swamp_height_above_sea_m
        && n < input.contract.swamp_noise_max
    {
        BiomeId::Swamp
    } else if distance_t > input.contract.plains_distance_min && n > input.contract.plains_noise_min
    {
        BiomeId::Plains
    } else if n > input.contract.forest_noise_min {
        BiomeId::Forest
    } else {
        BiomeId::Meadows
    };

    BiomeRegionSample {
        biome,
        region_noise: n,
        island_distance_t: distance_t,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn land_input(height: f32, noise_x: f32) -> BiomeRegionClassifyInput {
        BiomeRegionClassifyInput {
            x: noise_x,
            z: 0.0,
            height,
            seed: 0,
            sea_level: 18.0,
            region_cell_m: BIOME_REGION_CELL_M,
            island_radius_m: 560.0,
            island: IslandMaskSample {
                mask: 1.0,
                shore_distance_m: 200.0,
                nearest_center_x: 0.0,
                nearest_center_z: 0.0,
                cliff_weight: 0.0,
            },
            contract: BIOME_REGION_CONTRACT,
        }
    }

    #[test]
    fn ocean_and_coast_are_height_or_mask_driven() {
        let mut input = land_input(10.0, 0.0);
        assert_eq!(classify_biome_region(input).biome, BiomeId::Ocean);

        input.height = 18.5;
        assert_eq!(classify_biome_region(input).biome, BiomeId::Coast);

        input.height = 50.0;
        input.island.mask = 0.05;
        assert_eq!(classify_biome_region(input).biome, BiomeId::Ocean);
    }

    #[test]
    fn high_elevation_is_mountain() {
        assert_eq!(
            classify_biome_region(land_input(90.0, 0.0)).biome,
            BiomeId::Mountain
        );
    }

    #[test]
    fn contract_controls_classification_thresholds() {
        let default_input = land_input(90.0, 0.0);
        let mut overridden_input = default_input;
        overridden_input.contract = BiomeRegionContract {
            mountain_height_above_sea_m: 120.0,
            ..BIOME_REGION_CONTRACT
        };

        assert_eq!(
            classify_biome_region(default_input).biome,
            BiomeId::Mountain
        );
        assert_ne!(
            classify_biome_region(overridden_input).biome,
            BiomeId::Mountain
        );
    }

    #[test]
    fn field_sampling_is_deterministic() {
        let field = BiomeRegionField::default();
        let a = field.sample(512.0, -128.0, 72.0);
        let b = field.sample(512.0, -128.0, 72.0);
        assert_eq!(a, b);
    }

    #[test]
    #[should_panic(expected = "shared GPU contract")]
    fn region_cell_m_rejects_cpu_only_values() {
        let _ = BiomeRegionField::default().with_region_cell_m(BIOME_REGION_CELL_M + 1.0);
    }
}
